packages/hash/frontend/src/blocks/page/createPlaceholderPlugin/createPlaceholderPlugin.tsx

```diff
@@ -0,0 +1,96 @@
+import { paragraphBlockComponentId } from "@hashintel/hash-shared/blocks";
+import {
+  componentNodeToId,
+  findComponentNodes,
+} from "@hashintel/hash-shared/prosemirror";
+import { Schema } from "prosemirror-model";
+import { Plugin, PluginKey } from "prosemirror-state";
+import { Decoration, DecorationSet } from "prosemirror-view";
+import { RenderPortal } from "../usePortals";
+import { Placeholder } from "./Placeholder";
+
+interface PlaceholderPluginState {
+  focused: boolean;
+}
+
+const placeholderPluginKey = new PluginKey<PlaceholderPluginState, Schema>(
+  "placeholderPlugin",
+);
+
+export const createPlaceholderPlugin = (renderPortal: RenderPortal) => {
+  return new Plugin<PlaceholderPluginState, Schema>({
+    key: placeholderPluginKey,
+    state: {
+      init() {
+        return { focused: false };
+      },
+      apply(tr, state) {
+        const focused = tr.getMeta(placeholderPluginKey);
+
+        // return the old state if transaction does not have focused meta
+        if (focused === undefined) {
+          return state;
+        }
+
+        return { focused };
+      },
+    },
+    props: {
+      decorations(state) {
+        const firstNode = state.selection.$anchor.node(1);
+
+        if (!firstNode) return;
+
+        const componentId = componentNodeToId(
+          findComponentNodes(firstNode)[0]!,
+        );
+
+        const isFocused = placeholderPluginKey.getState(state)?.focused;
+        const isParagraph = componentId === paragraphBlockComponentId;
+        const isEmpty = findComponentNodes(firstNode)[0]?.childCount === 0;
+
+        const showPlaceholder = isParagraph && isEmpty && isFocused;
+
+        if (!showPlaceholder) return;
+
+        const widgetPos = state.selection.$anchor.posAtIndex(0, 1);
+
+        const placeholderDecoration = Decoration.widget(
+          widgetPos,
+          () => {
+            const mountNode = document.createElement("div");
+
+            renderPortal(<Placeholder />, mountNode);
+
+            return mountNode;
+          },
+          { key: "placeholder-deco" },
+        );
+
+        return DecorationSet.create(state.doc, [placeholderDecoration]);
+      },
+      handleDOMEvents: {
+        blur(view) {
+          view.dispatch(view.state.tr.setMeta(placeholderPluginKey, false));
+          return false;
+        },
+        focus(view) {
+          // this if prevents rendering `Placeholder` on readonly mode
+          if (!view.editable) return false;
```

Comment:
It's probably better to detect when this changes and track it in your plugin state, rather than ignoring changes to focus. What if the editor becomes editable *while* the editor is focused? (Can an editor be focused and not editable, I don't know)? It just seems more explicit.

Reply:
Editor can be focused while not-editable, which caused the placeholder to render on ready-only mode

Reply:
See [`f0c1ec1` (#980)](https://github.com/hashintel/hash/pull/980/commits/f0c1ec18865d1b1a6d4bce15e587c93050cdaf5d)

Reply:
Thanks, merged the suggested changes here https://github.com/hashintel/hash/pull/953/commits/02268173eb75cf43142cdc517b2f6dd510b1b66a